### PR TITLE
Add native Dataframe support for RLMs

### PR DESCRIPTION
**Pull request by kmad**
## Summary

This PR adds a `dspy.DataFrame` primitive which is intended for use by the RLM module, enabling LLMs to operate directly on DataFrames in the REPL with full type preservation. Non-RLM modules receive a string summary of the dataframe but will not be able to operate on its content. This PR also adds a `get_variable` function to allow for the extraction of variables from the REPL environment.

Disclosure: Code was produced with the assistance of Claude Opus 4.5 and reviewed manually.

### Core Features

**Parquet Serialization**
- DataFrames are serialized via PyArrow Parquet format and injected into the pyodide environment
- Preserves column dtypes: int64, float64, datetime64, categorical, boolean, etc.
- Binary data is base64-encoded for transport to the Pyodide sandbox

**Rich Metadata in Prompts**
- LLM receives DataFrame shape, memory usage, column dtypes, null counts
- First 3 and last 3 rows shown as preview
- Field descriptions and constraints from Signature are included

**`dspy.DataFrame` Type Annotation**
- New Pydantic-compatible type for use in DSPy Signatures
- Validates that values are pandas DataFrames
- Non-RLM modules (ChainOfThought, Predict) are shown a warning that only a string summary is provided

**New `get_variable` Method**
- Adds ability to retrieve variables from the sandbox after RLM execution
- Useful for extracting computed values not included in SUBMIT (e.g., base64-encoded charts)
- Returns JSON-deserialized values when possible, string representation otherwise

## Usage Example

```python
import dspy
import pandas as pd

class DataAnalysis(dspy.Signature):
    """Analyze the provided data."""
    data: dspy.DataFrame = dspy.InputField(desc="Sales data")
    summary: str = dspy.OutputField(desc="Analysis summary")
    total_revenue: float = dspy.OutputField()


# Configure your LM
lm = dspy.LM("anthropic/claude-sonnet-4-5-20250929", max_tokens=16000)
dspy.configure(lm=lm)

df = pd.DataFrame({
    "product": ["A", "B", "C"],
    "revenue": [100.0, 200.0, 150.0]
})

analyzer = dspy.RLM(DataAnalysis, max_iterations=10)
result = analyzer(data=df)

print(result.summary)
print(f"Total: ${result.total_revenue}")
```

## Files Changed

| File | Description |
|------|-------------|
| `dspy/primitives/repl_types.py` | Added `DataFrame` type with Pydantic schema support |
| `dspy/primitives/python_interpreter.py` | Added `get_variable`, DataFrame serialization, removed CDN access |
| `dspy/primitives/runner.js` | Added `get_variable` JSON-RPC handler, Parquet file injection |
| `dspy/primitives/__init__.py` | Export `DataFrame` |
| `docs/docs/api/primitives/DataFrame.md` | API documentation |
| `docs/docs/api/tools/PythonInterpreter.md` | Added `get_variable` to API docs |
| `docs/docs/tutorials/dataframes_rlm/index.ipynb` | Tutorial notebook |
| `docs/docs/tutorials/index.md` | Added tutorial link |

### `tests/primitives/test_python_interpreter.py`
- `get_variable` tests: string, list, dict, nested, computed, nonexistent, non-serializable
- DataFrame edge cases: empty, nulls, special column names, mixed dtypes
- DataFrame injection: basic, dtype preservation, operations, datetime, categorical, groupby

### `tests/primitives/test_repl_types.py`
- `dspy.DataFrame` type annotation in Signature
- Warning emission on serialization (verifies warning only triggers for non-RLM usage)
- Validation rejects non-DataFrame values
- Serialization to list of records

### `tests/predict/test_rlm.py`
- RLM with DataFrame input fields
- DataFrame column operations
- Dtype preservation through full pipeline
- `get_variable` retrieval after execution

## Test Results

```
tests/primitives/test_python_interpreter.py: 65 passed
tests/primitives/test_repl_types.py: 24 passed
tests/predict/test_rlm.py: 91 passed (excluding pre-existing async config issues)
```

### CODE REVIEW

**Comment by okhat** [comment]
So cool, thank you @kmad !!

File: dspy/primitives/python_interpreter.py [hunk 1]
```diff
@@ -551,6 +690,35 @@ def __call__(
...
+
+        Returns:
+            The variable's value (JSON-deserialized if possible, else string).
+
+        Raises:
+            CodeInterpreterError: If the variable doesn't exist or can't be retrieved.
+        """
+        self._check_thread_ownership()
+        self._ensure_deno_process()
+
```
[the remark this answers is not in the document]
Comment (kmad) [reply dspy/primitives/python_interpreter.py]: This has been resolved by adding validation using str.isidentifier() and  keyword.iskeyword() before sending to sandbox.

File: dspy/primitives/runner.js [hunk 2]
```diff
@@ -307,18 +307,51 @@ while (true) {
...
       try { pyodide.FS.mkdir('/tmp'); } catch (e) { /* exists */ }
       try { pyodide.FS.mkdir('/tmp/dspy_vars'); } catch (e) { /* exists */ }
-      pyodide.FS.writeFile(`/tmp/dspy_vars/${name}.json`, new TextEncoder().encode(value));
+
+      if (format === "parquet") {
+        // Pre-load pyarrow for parquet support (pandas needs it for read_parquet)
+        await pyodide.loadPackage(["pandas", "pyarrow"]);
+        // Decode base64 and write binary Parquet data
+        const binaryData = Uint8Array.from(atob(value), c => c.charCodeAt(0));
+        pyodide.FS.writeFile(`/tmp/dspy_vars/${name}.parquet`, binaryData);
```
[the remark this answers is not in the document]
Comment (kmad) [reply dspy/primitives/runner.js]: Great callout. Revised to simply call Deno's decodeBase64 (included by default) to prevent additional string copies.

File: dspy/primitives/repl_types.py [hunk 3]
```diff
@@ -78,6 +146,76 @@ def from_value(
...
+            f"Memory: {memory_str}",
+            "",
+            "Columns:",
+            *columns_info,
+            "",
+            "First 3 rows:",
+            df.head(3).to_string(),
+        ]
+
+        if shape[0] > 6:
```
[the remark this answers is not in the document]
Comment (kmad) [reply dspy/primitives/repl_types.py:138]: Revised to implement preview_chars properly.

File: docs/docs/api/primitives/DataFrame.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+```
+
+## Retrieving Variables After Execution
+
+After RLM execution, you can retrieve variables from the sandbox using `get_variable()`:
+
+```python
+# Access the interpreter after execution
+interpreter = result._interpreter
+
```
[the remark this answers is not in the document]
Comment (kmad) [reply docs/docs/api/primitives/DataFrame.md]: Updated documentation to show use of optional `PythonInterpreter` definition to use with `get_variable`. Fixed exposure of _interpreter object.

File: dspy/primitives/repl_types.py [hunk 5]
```diff
@@ -10,17 +10,81 @@
...
+        cls, source_type: Any, handler: GetCoreSchemaHandler
+    ) -> CoreSchema:
+        """Return a Pydantic core schema that accepts any DataFrame instance."""
+        import warnings
+
+        warnings.warn(
+            "dspy.DataFrame should only be used with dspy.RLM. "
+            "Other modules (ChainOfThought, Predict, etc.) will only see a string "
+            "representation, not the actual DataFrame data.",
+            UserWarning,
```
[the remark this answers is not in the document]
Comment (kmad) [reply dspy/primitives/repl_types.py]: Moved warning to `serialize_dataframe` function instead of global pydantic core schema. Serialization should only occur when using non-RLM Modules.

File: docs/docs/tutorials/dataframes_rlm/index.ipynb [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
...Rlyd3dXQECAU7vNZlNwcLAyMjKc2hs2bOj03NPTs8L28+M5v0xh6tSp8vDwcHqMGzdOknTixImKT/oPHD9+XFlZWfL09CxzjNTUVKf+R48erTNnzmjp0qWSpM8//1wpKSlOyy6OHz8uwzAUFBRUpr9t27Zd8ngBXJ7cXT0AAIAUEBCgTZs2qbi4uNyw7Ofnp7Nnzyo9Pd0pLBuGodTUVHXs2LFKxuLv7y9Jmj59uoYMGWJZExERccnH8PPzU0JCguV2Hx8f8+uWLVvqhhtu0IIFCzR27FgtWLBAoaGh6tOnj1N/NptNX3/9tex2e5n+rNoA4I8wowwAtUD//v115syZCt8o16tXL0nSO++849S+bNky5ebmmtsvVUREhJo2barvvvtOHTp0sHyUDLIXIyoqShkZGSoqKrLsv3QQHzVqlLZv365Nmzbp008/1ciRI+Xm5ubUn2EY+u233yz7a9269SWNF8DliRllAKgFRowYoQULFuj+++/X/v371bNnTxUXF2v79u1q0aKFhg8frltvvVV9+/bVo48+qpycHHXr1k3ff/+9Zs6cqbZt2yo2NrbKxjN//nz1799fffv2VVxcnK688kqdPHlS+/bt086dO/Xhhx/+YR8///yz/vd//7dMe8uWLTV8+HC9++67uu222/TQQw/phhtukIeHh44ePar169dr0KBBuv32252uz+TJkzVixAjl5+eXWafdrVs33XfffRo1apR27Nihm2++Wd7e3kpJSdGmTZvUunVrPfDAA5d8XQBcXgjKAFALuLu767PPPtNTTz2l9957Ty+++KJ8fHzUpk0b9evXT9K5tcgff/yxZs2apQULFujJJ5+Uv7+/YmNjNWfOnCpdXtCzZ0998803evLJJzVp0iRlZmbKz89PLVu21NChQyvVR0JCguXSipkzZ2rWrFlasWKFXnrpJS1evFhPPfWU3N3d1ahRI3Xv3r3MDLDD4dDtt9+uJUuWqFu3bmrWrFmZfufPn6/OnTtr/vz5evXVV1VcXKzQ0FB169ZNN9xww8VdCACXNZthGIarBwEAAADUNqxRBgAAACwQlAEAAAALBGUAAADAAkEZAAAAsEBQBgAAACwQlAEAAAALBGUAAADAAkEZAAAAsEBQBgAAACwQlAEAAAALBGUAAADAwv8Hu1wLZzcUl8UAAAAASUVORK5CYII=\"/>"
```
[the remark this answers is not in the document]
Comment (kmad) [reply docs/docs/tutorials/dataframes_rlm/index.ipynb]: Cleared notebook.

File: tests/primitives/test_repl_types.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    def test_dataframe_type_warns_on_use(self):
+        """Test warning is raised when dspy.DataFrame is used in Signature."""
+        import warnings
+        import dspy
+        from dspy.primitives.repl_types import DataFrame
+
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+
+            class WarningSig(dspy.Signature):
```
[the remark this answers is not in the document]
Comment (kmad) [reply tests/primitives/test_repl_types.py]: Removed unused class, as part of move to `serialize_dataframe`

File: dspy/primitives/runner.js [hunk 8]
```diff
@@ -307,18 +307,51 @@ while (true) {
...
 
+  if (method === "get_variable") {
+    const { name } = params;
+    try {
+      // Get the variable from Python globals and serialize it
+      const result = pyodide.runPython(`
+import json
+_var = ${name}
+if hasattr(_var, 'to_dict'):
+    _result = json.dumps(_var.to_dict())
```
[the remark this answers is not in the document]
Comment (kmad) [reply dspy/primitives/runner.js]: Resolved by adding additional checks and properly accessing the variable.

**Comment by kmad** [comment]
Copilot had great suggestions. They have all been resolved in the latest commits.

File: docs/docs/tutorials/dataframes_rlm/index.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (okhat) [reply docs/docs/tutorials/dataframes_rlm/index.ipynb]: Thanks so much @kmad ! I don't want you to do so many unnecessary tasks before we merge (just waiting for @isaacbmiller 's stamp), but I feel like I might disagree with Copilot on this one. It's very nice that the tutorial has the expected outputs in a readable format on the website... What do you think?

[the remark this answers is not in the document]
Comment (kmad) [reply docs/docs/tutorials/dataframes_rlm/index.ipynb]: I've generally found notebook output to be useful for skimming expected functionality (assuming it's not a bunch of messy progress bar output and the like). The output here is pretty concise so happy to add it back in.

**Comment by kmad** [comment]
I added the notebook examples back in per Omar's comment. @isaacbmiller let me know if you have any other suggestions.

File: dspy/primitives/python_interpreter.py [hunk 9]
```diff
@@ -167,10 +185,20 @@ def __init__(
...
+            # Network access: user-specified domains
             if self.enable_network_access:
-                args.append(f"--allow-net={','.join(str(x) for x in self.enable_network_access)}")
+                allowed_net = [str(x) for x in self.enable_network_access]
+                args.append(f"--allow-net={','.join(allowed_net)}")
+
+            # Write paths: Deno cache (for Pyodide package caching) + user-specified paths
+            allowed_write_paths = []
+            if deno_dir:
```
Comment (isaacbmiller) [inline dspy/primitives/python_interpreter.py]: Why do we need to allow deno_dir in the write paths?

File: dspy/primitives/repl_types.py [hunk 10]
```diff
@@ -10,17 +10,83 @@
...
+
+# Type alias for use in DSPy Signatures: `dataframe: dspy.DataFrame = dspy.InputField()`
+# This uses Annotated to attach the Pydantic schema handler to the pandas DataFrame type
+#
+# WARNING: dspy.DataFrame should only be used with dspy.RLM, which provides a Python
+# sandbox where the DataFrame is available for code execution. Other modules
+# (ChainOfThought, Predict, etc.) will only see a string representation.
+try:
+    import pandas as pd
+    DataFrame = Annotated[pd.DataFrame, _DataFrameAnnotation()]
```
Comment (isaacbmiller) [inline dspy/primitives/repl_types.py]: Prefer to use `dspy.Type` over Annotated if possible

File: dspy/primitives/runner.js [hunk 11]
```diff
@@ -307,18 +308,60 @@ while (true) {
...
+      }
+
       console.log(jsonrpcResult({ injected: name }, requestId));
     } catch (e) {
       console.log(jsonrpcError(JSONRPC_APP_ERRORS.RuntimeError, `Failed to inject var: ${e.message}`, requestId));
     }
     continue;
   }
 
+  if (method === "get_variable") {
```
Comment (isaacbmiller) [inline dspy/primitives/runner.js]: Why do we need this method?

File: dspy/primitives/repl_types.py (hunk 10 above)
Comment (kmad) [reply dspy/primitives/repl_types.py]: Fair point; this is now a proper `Type`. Let me know if this implementation is satisfactory.

File: dspy/primitives/python_interpreter.py (hunk 9 above)
Comment (kmad) [reply dspy/primitives/python_interpreter.py]: Good catch, thanks. This was required when loading external packages but pandas/pyarrow are built in in the latest version of pyodide. This is not needed, and has been removed.

File: dspy/primitives/runner.js (hunk 11 above)
Comment (kmad) [reply dspy/primitives/runner.js]: I thought this might be helpful for extracting variables/ objects modified by the RLM. It's admittedly a separate feature/ capability from the `Dataframe` implementation, but could be useful when used together. E.g., I could see a directive of "given this big dataframe, perform X,Y,Z transformations on it" - and then get back the dataframe (or other derivative variables) symbolically. But up to you if it's creating too much surface area for this PR, happy to move somewhere else.

Comment (isaacbmiller) [reply dspy/primitives/runner.js]: I would rather the return type of your signature be a Dataframe than build in this function specifically. I do see the value, but interacting with the REPL outside of the bounds of a signature is an antipattern IMO.

Let's remove

File: tests/predict/test_rlm.py [hunk 12]
```diff
@@ -1049,5 +1049,137 @@ def test_with_llm_query(self):
...
+
+        rlm = RLM(SumColumn, max_iterations=3)
+        rlm.generate_action = make_mock_predictor([
+            {"reasoning": "Sum the value column", "code": "SUBMIT(int(df['value'].sum()))"},
+        ])
+
+        result = rlm.forward(df=df)
+        assert result.total == 100
+
+    def test_dataframe_with_string_signature(self):
```
Comment (isaacbmiller) [inline tests/predict/test_rlm.py]: slop

File: tests/predict/test_rlm.py [hunk 13]
```diff
@@ -1049,5 +1049,137 @@ def test_with_llm_query(self):
...
+class TestRLMWithDataFrame:
+    """Integration tests for RLM with DataFrame inputs.
+
+    These tests verify the full pipeline: DataFrame -> Parquet serialization
+    -> Pyodide sandbox -> pandas operations -> result.
+    """
+
+    def test_dataframe_input_basic(self):
+        """Test RLM with DataFrame input field."""
+        import warnings
```
Comment (isaacbmiller) [inline tests/predict/test_rlm.py]: move all of the repeated imports to the top of the file.

File: tests/predict/test_rlm.py [hunk 14]
```diff
@@ -1049,5 +1049,137 @@ def test_with_llm_query(self):
...
+
+        import dspy
+        from dspy.primitives.repl_types import DataFrame
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", UserWarning)
```
Comment (isaacbmiller) [inline tests/predict/test_rlm.py]: What is the purpose of filtering this warning?

File: dspy/primitives/runner.js (hunk 11 above)
Comment (kmad) [reply dspy/primitives/runner.js]: Fair enough, agree on keeping the abstraction boundaries clean. Removed in latest commit.

File: tests/predict/test_rlm.py (hunk 12 above)
Comment (kmad) [reply tests/predict/test_rlm.py]: Removed unnecessary test.

File: tests/predict/test_rlm.py (hunk 13 above)
Comment (kmad) [reply tests/predict/test_rlm.py]: Done.

File: tests/predict/test_rlm.py (hunk 14 above)
Comment (kmad) [reply tests/predict/test_rlm.py]: These warnings were leftover from the Annotation approach; removed now. The purpose of the warnings more broadly is to notify the developer if they are serializing the dataframe - to be aware that the LLM may no longer have symbolic access to the data.

**Comment by isaacbmiller** [comment]
Hi @kmad,

I have been mulling over this PR over the weekend, and I do not think that we want to merge support for a type that is only useful for a specific module and changes to the hot path of the RLm modules to support that.

The PR I would rather see is:
1. How do we support a user wanting to pass in and (optionally) get an arbitrary type from, an RLM
This notably starts to look like an adapter. I will be designing some interfaces that this can work for in the next few weeks, but I would greatly appreciate help with figuring out the right DX!
2. For this generic interface, here is how you can add support for any type, e.g. DataFrames, in a tutorial notebook.

This is especially hard as different types require imports in both the main DSPy thread and in the pyodide sandbox running underneath.

Thank you for the help!

**Comment by darinkishore** [comment]
@isaacbmiller 

i'm not sure how this will be possible in the pyodide sandbox; I also agree this direction is important. 

you may want to take a look at the current WIP implementation in
 https://github.com/krypticmouse/DSRs/tree/wip/rlm-storage-20260130 . 
 
 i don't think the details (especially for the variable helper previews) are all fully there, but it is 80% there and it works. users can pass in arbitary rust types and get idiomatic python types, model gets to work with native python, you can add in your own methods, the works.

**Comment by kmad** [comment]
That makes sense - I agree the initial implementation is too Dataframe specific and breaks assumptions about how things should be interchangable. Will open a new PR with generic RLM Type support with Dataframe implementation.
